### PR TITLE
xcp ref implementation

### PR DESCRIPTION
**Pull request by DhanusML**
## Notation
Data is a matrix $X\in\mathbb{R}^{p\times n}$. Each column is a $p$-dimensional vector sampled independently. The matrix  $X$ is assumed to be stored in column-major fashion.

##  xcp
Matrix of cross products is a $p\times p$ matrix whose $(i,j)$th entry is given by
$$C_{ij} = \sum_{k=1}^n(x_{ik}-\mu_i)(x_{jk}-\mu_j).$$
Here $x_{ik}$ is the $i$th component of the $k$th random vector.
 
The implementation requires this to be computed in batches of $X$, where the raw sum is passed between the batch computations. This can be done using the following:

Let $\mu'$ be the mean of the data until the previous batch and let $\mu$ be the mean including the data from current batch. Let $n'$ be the total number of data points until the previous batch and $S'$ be their sum (i.e., $\mu' = S'/n'$ and $\mu = S/n$).

Then the $(i,j)$ entry of $C$ can be computed using
$$C_{ij}\leftarrow C_{ij}+(\mu_j'-\mu_j)S_i' + (\mu_i'-\mu_i)S_j' + (\mu_i\mu_j-\mu_i'\mu_j')n' + \sum_{k=n'+1}^n(x_{ik}-\mu_i)(x_{jk}-\mu_j).$$ 

This can be simplified to
$$C_{ij} \leftarrow C_{ij} + \frac{S_i' S_j'}{n'} - \frac{S_i S_j}{n} + \sum_{k=n'+1}^n x_{ik}x_{jk}$$
or
$$C \leftarrow C + \frac{S'S'^t}{n'} - \frac{SS^t}{n} + XX^t$$

The level3 BLAS routine GEMM is used to compute the matrix products.

### CODE REVIEW

File: cpp/daal/src/externals/service_stat_ref.h [hunk 1]
```diff
@@ -174,6 +175,56 @@ struct RefStatistics<double, cpu>
...
+        double accWtOld = *nPreviousObservations;
+        double accWt    = *nPreviousObservations + nVectors;
+        DAAL_INT one    = 1;
+        char transa     = 'N';
+        char transb     = 'N';
+        double beta     = 0.0;
+        double alpha;
+        if (accWtOld != 0)
+        {
+            double * sumOld = daal::services::internal::service_malloc<double, cpu>(nFeatures, sizeof(double));
```
Comment (KulikovNikita) [inline cpp/daal/src/externals/service_stat_ref.h:188]:
```suggestion
            double* const sumOld = daal::services::internal::service_malloc<double, cpu>(nFeatures, sizeof(double));
```

File: cpp/daal/src/externals/service_stat_ref.h [hunk 2]
```diff
@@ -174,6 +175,56 @@ struct RefStatistics<double, cpu>
...
+            DAAL_CHECK_MALLOC(sumOld);
+            for (DAAL_INT i = 0; i < nFeatures; ++i)
+            {
+                sumOld[i] = sum[i];
+            }
+            // S_old S_old^t/accWtOld
+            alpha = 1.0 / accWtOld;
```
Comment (KulikovNikita) [inline cpp/daal/src/externals/service_stat_ref.h:195]: Any checks for overflow?

File: cpp/daal/src/externals/service_stat_ref.h [hunk 3]
```diff
@@ -174,6 +175,56 @@ struct RefStatistics<double, cpu>
...
+        }
+
+        // -S S^t/accWt
+        alpha = -1.0 / accWt;
+        blasInst.xgemm(&transa, &transb, &nFeatures, &nFeatures, &one, &alpha, sum, &nFeatures, sum, &one, &beta, crossProduct, &nFeatures);
+
+        // X X^t
+        transb = 'T';
+        alpha  = 1.0;
+        beta   = 1.0;
```
Comment (KulikovNikita) [inline cpp/daal/src/externals/service_stat_ref.h:225]: I would recommend not to use the same variables. It's confusing and also can be dangerous if someone will forgot to change them.

File: cpp/daal/src/externals/service_stat_ref.h [hunk 4]
```diff
@@ -174,6 +175,56 @@ struct RefStatistics<double, cpu>
...
+        blasInst.xgemm(&transa, &transb, &nFeatures, &nFeatures, &nVectors, &alpha, data, &nFeatures, data, &nFeatures, &beta, crossProduct,
+                       &nFeatures);
```
Comment (KulikovNikita) [inline cpp/daal/src/externals/service_stat_ref.h:227]:
```suggestion
        {
            constexpr char transb = 'T';
            constexpr double alpha  = 1.0;
            constexpr double beta   = 1.0;
            blasInst.xgemm(&transa, &transb, &nFeatures, &nFeatures, &nVectors, &alpha, data, &nFeatures, data, &nFeatures, &beta, crossProduct,
                       &nFeatures);
        }
```

File: cpp/daal/src/externals/service_stat_ref.h (hunk 2 above)
Comment (DhanusML) [reply cpp/daal/src/externals/service_stat_ref.h]: Does onedal have some macros to check floating point overflow?

**Comment by Alexsandruss** [comment]
@KulikovNikita, were your comments addressed?

**Comment by DhanusML** [comment]
Can you provide details of the check that is failing?

**Comment by Alexsandruss** [comment]
> Can you provide details of the check that is failing?

I suppose they are sporadic failures. Restarting CI

**Review by Alexandr-Solovev — APPROVED** [review]

**Review by Alexsandruss — APPROVED** [review]

**Pull request merged**